dandi/files/zarr.py

```diff
@@ -0,0 +1,591 @@
+from __future__ import annotations
+
+import atexit
+from collections.abc import Generator, Iterator
+from concurrent.futures import Future, ThreadPoolExecutor, as_completed
+from contextlib import closing
+from dataclasses import dataclass, field, replace
+from datetime import datetime
+import os
+from pathlib import Path
+from time import sleep
+from typing import Any, Optional, cast
+
+from dandischema.digests.zarr import get_checksum
+from dandischema.models import BareAsset, DigestType
+import requests
+import zarr
+
+from dandi import get_logger
+from dandi.consts import (
+    MAX_ZARR_DEPTH,
+    ZARR_MIME_TYPE,
+    ZARR_UPLOAD_BATCH_SIZE,
+    EmbargoStatus,
+)
+from dandi.dandiapi import (
+    RemoteAsset,
+    RemoteDandiset,
+    RemoteZarrAsset,
+    RemoteZarrEntry,
+    RESTFullAPIClient,
+)
+from dandi.metadata import get_default_metadata
+from dandi.misctypes import BasePath, Digest
+from dandi.support.digests import get_digest, get_zarr_checksum, md5file_nocache
+from dandi.utils import chunked, pluralize
+
+from .bases import LocalDirectoryAsset
+
+lgr = get_logger()
+
+
+@dataclass
+class LocalZarrEntry(BasePath):
+    """A file or directory within a `ZarrAsset`"""
+
+    #: The path to the actual file or directory on disk
+    filepath: Path
+    #: The path to the root of the Zarr file tree
+    zarr_basepath: Path
+
+    def _get_subpath(self, name: str) -> LocalZarrEntry:
+        if not name or "/" in name:
+            raise ValueError(f"Invalid path component: {name!r}")
+        elif name == ".":
+            return self
+        elif name == "..":
+            return self.parent
+        else:
+            return replace(
+                self, filepath=self.filepath / name, parts=self.parts + (name,)
+            )
+
+    @property
+    def parent(self) -> LocalZarrEntry:
+        if self.is_root():
+            return self
+        else:
+            return replace(self, filepath=self.filepath.parent, parts=self.parts[:-1])
+
+    def exists(self) -> bool:
+        return self.filepath.exists()
+
+    def is_file(self) -> bool:
+        return self.filepath.is_file()
+
+    def is_dir(self) -> bool:
+        return self.filepath.is_dir()
+
+    def iterdir(self) -> Iterator[LocalZarrEntry]:
+        for p in self.filepath.iterdir():
+            if p.is_dir() and not any(p.iterdir()):
+                # Ignore empty directories
+                continue
+            yield self._get_subpath(p.name)
+
+    def get_digest(self) -> Digest:
+        """
+        Calculate the DANDI etag digest for the entry.  If the entry is a
+        directory, the algorithm will be the Dandi Zarr checksum algorithm; if
+        it is a file, it will be MD5.
+        """
+        if self.is_dir():
+            return Digest.dandi_zarr(get_zarr_checksum(self.filepath))
+        else:
+            return Digest(
+                algorithm=DigestType.md5, value=get_digest(self.filepath, "md5")
+            )
```

Comment:
I think we have unittests for digesting zarrs, please unittest this as well.

Reply:
Tested: 7d07424